src/data/bucket/circle_bucket.js

```diff
@@ -70,20 +73,33 @@ class CircleBucket<Layer: CircleStyleLayer | HeatmapStyleLayer> implements Bucke
         for (const {feature, index, sourceLayerIndex} of features) {
             if (this.layers[0]._featureFilter({zoom: this.zoom}, feature)) {
                 const geometry = loadGeometry(feature);
-                this.addFeature(feature, geometry);
+                this.addFeature(feature, geometry, index);
                 options.featureIndex.insert(feature, geometry, index, sourceLayerIndex, this.index);
             }
         }
     }
 
+    update(states: FeatureStates, vtLayer: VectorTileLayer) {
+        if (!this.stateDependentLayers.length) return;
+        this.changed = this.programConfigurations.updatePaintArrays(states, vtLayer, this.stateDependentLayers);
```

Comment:
> If we follow that idea, the correct place to track changes would be in the individual ProgramConfiguration's

We can put a `dirty` property on `ProgramConfigurationSet`, which is set during `updatePaintArrays`, checked in `upload()` to decide whether to actually do work (and then cleared).  I suppose it might not be simpler per se, but it has the advantage of localizing and centralizing this state tracking in `program_configuration.js`, rather than having it replicated in each bucket.